ffi/derive/src/convert.rs

```diff
@@ -431,472 +488,380 @@ fn derive_try_from_repr_c_for_fieldless_enum(
-
-    let (impl_generics, ty_generics, where_clause) = &input.generics.split_for_impl();
 
-    quote! {
-        impl #impl_generics iroha_ffi::IntoFfi for #name #ty_generics #where_clause {
-            type Target = *mut Self;
+    fn gen_repr_c_enum_name(enum_name: &Ident) -> Ident {
+        Ident::new(
+            &format!("__iroha_ffi__ReprC{}", enum_name),
+            proc_macro2::Span::call_site(),
+        )
+    }
 
-            fn into_ffi(self) -> Self::Target {
-                let layout = core::alloc::Layout::for_value(&self);
+    fn gen_repr_c_enum_payload_name(enum_name: &Ident) -> Ident {
+        Ident::new(
+            &format!("__iroha_ffi__{}Payload", enum_name),
+            proc_macro2::Span::call_site(),
+        )
+    }
 
-                unsafe {
-                    let ptr: Self::Target = alloc(layout).cast();
-                    ptr.write(self);
-                    ptr
-                }
+    fn variant_mapper<F0: FnOnce() -> TokenStream2, F1: FnOnce(&syn::Field) -> TokenStream2>(
+        variant: &syn::Variant,
+        unit_mapper: F0,
+        field_mapper: F1,
+    ) -> TokenStream2 {
+        match &variant.fields {
+            syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }) if unnamed.len() == 1 => {
+                field_mapper(&unnamed[0])
             }
-        }
-
-        impl #impl_generics iroha_ffi::IntoFfi for &#name #ty_generics #where_clause {
-            type Target = *const #name #ty_generics;
-
-            fn into_ffi(self) -> Self::Target {
-                <*const _>::from(self)
+            syn::Fields::Unnamed(syn::FieldsUnnamed { unnamed, .. }) => {
+                abort!(unnamed, "Only 1-sized variants are supported")
             }
-        }
-
-        impl #impl_generics iroha_ffi::IntoFfi for &mut #name #ty_generics #where_clause {
-            type Target = *mut #name #ty_generics;
-
-            fn into_ffi(self) -> Self::Target {
-                <*mut _>::from(self)
+            syn::Fields::Named(syn::FieldsNamed { named, .. }) => {
+                abort!(named, "Named variants are not supported")
             }
+            syn::Fields::Unit => unit_mapper(),
         }
-
-        #opaque_item_slice_into_ffi_derive
-        #opaque_item_vec_into_ffi_derive
     }
 }
 
-fn derive_into_ffi_for_opaque_item_slice(input: &DeriveInput) -> TokenStream2 {
+fn derive_transparency(input: &syn::DeriveInput) -> TokenStream2 {
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
     let name = &input.ident;
-    let mut generics = input.generics.clone();
-    let lifetime: syn::Lifetime = syn::parse_quote!('__iroha_ffi_slice);
-    let lifetimes = &[lifetime.clone()];
-    let (generic_params, ty_generics, where_clause) =
-        add_bounds_and_split_generics(&mut generics, lifetimes, &[], &[]);
+    let inner = match &input.data {
+        // TODO: Do we check it's not ZST?
```

Comment:
we could add such check, but it would produce akward errors in case if `inner` is generic.

```
const _: () = assert!(core::mem::size_of::<#inner>() > 0, "inner field is ZST");
```

Reply:
we should make an issue to track this